### PR TITLE
Build RAPIDS.

### PR DESCRIPTION
**Pull request by bdice**
It'd be nice to have nightly builds of RAPIDS in devcontainers, as a way to test end-to-end builds and pre-populate sccache.

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -145,3 +145,12 @@ jobs:
...
       repo: "${{ vars.DOCKERHUB_REPOSITORY || github.repository }}"
+
+  build-rapids:
+    secrets: inherit
+    uses: rapidsai/shared-action-workflows/.github/workflows/build-in-devcontainer.yaml@branch-24.02
+    with:
+      build_command: |
+        sccache -z;
+        build-all --verbose;
+        sccache -s;
```
Comment (trxcllnt) [inline .github/workflows/test.yml:156]:
```suggestion
        sccache -z;
        clone-all --verbose;
        build-all --verbose;
        sccache -s;
```

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -145,3 +145,12 @@ jobs:
       features: "${{ toJSON(matrix.features) }}"
       container_env: "${{ toJSON(matrix.env) }}"
       repo: "${{ vars.DOCKERHUB_REPOSITORY || github.repository }}"
+
+  build-rapids:
```
Comment (trxcllnt) [inline .github/workflows/test.yml:149]:
```suggestion
  build-rapids:
    if: needs.check-event.outputs.ok == 'true'
```

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -145,3 +145,20 @@ jobs:
...
+    if: needs.check-event.outputs.ok == 'true'
```
Comment (trxcllnt) [inline .github/workflows/test.yml:150]:
```suggestion
    if: needs.check-event.outputs.ok == 'true'
    needs: check-event
```

Comment (bdice) [inline .github/workflows/test.yml:150]:
```suggestion
    if: needs.check-event.outputs.ok == 'true'
    needs: check-event
```

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -145,3 +145,21 @@ jobs:
...
+      build_command: |
+        sccache -z;
+        clone-all --verbose;
+        build-all --verbose \
+          -DBUILD_TESTS=ON \
+          -DBUILD_BENCHMARKS=ON \
+           -DBUILD_ANN_BENCH=ON \
+           -DBUILD_PRIMS_BENCH=ON \
+          -DNVBench_ENABLE_CUPTI=OFF \
+           -j$(nproc --ignore=1);
```
Comment (trxcllnt) [inline .github/workflows/test.yml:164]:
```suggestion
        build-all --verbose \
          -DBUILD_TESTS=ON \
          -DBUILD_BENCHMARKS=ON \
          -DBUILD_ANN_BENCH=ON \
          -DBUILD_PRIMS_BENCH=ON \
          -j$(nproc --ignore=1);
```

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -145,3 +145,24 @@ jobs:
...
+      build_command: |
+        sccache -z;
+        clone-all --verbose --no-update-env;
+        rapids-update-content-command;
+        rapids-post-attach-command;
+        build-all --verbose \
+          -DBUILD_TESTS=ON \
+          -DBUILD_BENCHMARKS=ON \
+          -DBUILD_ANN_BENCH=ON \
+          -DBUILD_PRIMS_BENCH=ON \
```
Comment (trxcllnt) [inline .github/workflows/test.yml:166]:
```suggestion
          -DBUILD_PRIMS_BENCH=ON \
          $([ "pip" = $PYTHON_PACKAGE_MANAGER ] && echo -DUSE_LIBARROW_FROM_PYARROW=ON || echo)
```

**Comment by bdice** [comment]
Nice! The conda job passed in 186 minutes. Let's see if subsequent runs are faster (they should be much faster with sccache).

**Comment by trxcllnt** [comment]
Blocked by ~https://github.com/rapidsai/ucxx/pull/169~ and https://github.com/rapidsai/wholegraph/pull/118

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -145,3 +145,24 @@ jobs:
...
+    secrets: inherit
+    uses: rapidsai/shared-action-workflows/.github/workflows/build-in-devcontainer.yaml@fix/devcontainer-json-location
+    with:
+      node_type: cpu32
+      gh_token: GIST_REPO_READ_ORG_GITHUB_TOKEN
+      build_command: |
+        sccache -z;
+        clone-all --verbose --no-update-env;
+        rapids-update-content-command;
+        rapids-post-attach-command;
```
Comment (trxcllnt) [inline .github/workflows/test.yml:161]:
```suggestion
        rapids-make-${PYTHON_PACKAGE_MANAGER}-env;
```

**Comment by trxcllnt** [comment]
Seems like `ucxx` is trying to install `doxygen==1.10.0`

File: .github/workflows/test.yml [hunk 7]
```diff
@@ -145,3 +145,25 @@ jobs:
...
+  build-rapids:
+    if: needs.check-event.outputs.ok == 'true'
+    needs: check-event
+    secrets: inherit
+    uses: rapidsai/shared-action-workflows/.github/workflows/build-in-devcontainer.yaml@fix/devcontainer-json-location
+    with:
+      node_type: cpu32
+      gh_token: GIST_REPO_READ_ORG_GITHUB_TOKEN
+      build_command: |
+        set -Eeuo pipefail;
```
Comment (trxcllnt) [inline .github/workflows/test.yml:158]:
```suggestion
        set -Eeuxo pipefail;
```

File: .github/workflows/test.yml [hunk 8]
```diff
@@ -145,3 +145,25 @@ jobs:
...
+        sccache -z;
+        clone-all --verbose --no-update-env;
+        rapids-make-${PYTHON_PACKAGE_MANAGER}-env;
+        build-all --verbose \
+          -DBUILD_TESTS=ON \
+          -DBUILD_BENCHMARKS=ON \
+          -DBUILD_ANN_BENCH=ON \
+          -DBUILD_PRIMS_BENCH=ON \
+          $([ "pip" = $PYTHON_PACKAGE_MANAGER ] && echo -DUSE_LIBARROW_FROM_PYARROW=ON || echo) \
+          -j$(nproc --ignore=1);
```
Comment (trxcllnt) [inline .github/workflows/test.yml:168]:
```suggestion
        rapids_build_utils_debug=all \
        build-all \
          --verbose \
          -j$(nproc --ignore=1) \
          -DBUILD_TESTS=ON \
          -DBUILD_BENCHMARKS=ON \
          -DBUILD_ANN_BENCH=ON \
          -DBUILD_PRIMS_BENCH=ON \
          $([ "pip" = ${PYTHON_PACKAGE_MANAGER} ] && echo -DUSE_LIBARROW_FROM_PYARROW=ON || echo) \
          ;
```

File: .github/workflows/test.yml [hunk 9]
```diff
@@ -145,3 +145,28 @@ jobs:
...
+    uses: rapidsai/shared-action-workflows/.github/workflows/build-in-devcontainer.yaml@fix/devcontainer-json-location
+    with:
+      node_type: cpu32
+      gh_token: GIST_REPO_READ_ORG_GITHUB_TOKEN
+      build_command: |
+        set -Eeuxo pipefail;
+        sccache -z;
+        clone-all --verbose --no-update-env;
+        rapids-make-${PYTHON_PACKAGE_MANAGER}-env;
+        rapids_build_utils_debug=all \
```
Comment (trxcllnt) [inline .github/workflows/test.yml:162]:
```suggestion
```

File: .github/workflows/test.yml [hunk 10]
```diff
@@ -145,3 +145,28 @@ jobs:
       features: "${{ toJSON(matrix.features) }}"
       container_env: "${{ toJSON(matrix.env) }}"
       repo: "${{ vars.DOCKERHUB_REPOSITORY || github.repository }}"
+
+  build-rapids:
+    if: needs.check-event.outputs.ok == 'true'
```
Comment (trxcllnt) [inline .github/workflows/test.yml:150]:
```suggestion
    name: Build all RAPIDS repos
    if: needs.check-event.outputs.ok == 'true'
```

**Comment by trxcllnt** [comment]
@bdice the ucxx PR needed to target branch-0.36 to make this build succeed

**Comment by bdice** [comment]
Sorry. I was focused on getting that PR aligned with a few others for wheels and CUDA 12.2 support that were targeting 0.37 (24.04). Can we retarget this PR for 24.04?

File: .github/workflows/test.yml [hunk 11]
```diff
@@ -145,3 +145,28 @@ jobs:
...
+    uses: rapidsai/shared-action-workflows/.github/workflows/build-in-devcontainer.yaml@fix/devcontainer-json-location
+    with:
+      node_type: cpu32
+      gh_token: GIST_REPO_READ_ORG_GITHUB_TOKEN
+      build_command: |
+        sccache -z;
+        clone-all --verbose --no-update-env;
+        rapids-generate-scripts;
+        rapids-make-${PYTHON_PACKAGE_MANAGER}-env;
+        build-all \
```
Comment (trxcllnt) [inline .github/workflows/test.yml:163]:
```suggestion
        set -e;
        build-all \
```

**Comment by bdice** [comment]
This forward-merger for cuml should unblock the current build failure: https://github.com/rapidsai/cuml/pull/5735

**Comment by trxcllnt** [comment]
It's failing because RAFT is built independently of cuML. We need this fix in RAFT:
```cmake
    # in get_hnswlib.cmake
    rapids_export_find_package_root(
      BUILD hnswlib [=[${CMAKE_CURRENT_LIST_DIR}]=] EXPORT_SET raft-exports
    )
```

**Comment by trxcllnt** [comment]
CI is green, so I'm merging this. Thanks @bdice!

**Comment by trxcllnt** [comment]
Last thing to do is verify the [`fix/devcontainer-json-location`](https://github.com/rapidsai/shared-workflows/tree/fix/devcontainer-json-location) branch doesn't break other RAPIDS libs, then merge those changes into `rapidsai/shared-workflows@branch-24.04`.

**Comment by trxcllnt** [comment]
We will also need ops to add the `CUMLPRIMS_SSH_PRIVATE_DEPLOY_KEY` and `CUGRAPH_OPS_SSH_PRIVATE_DEPLOY_KEY` secrets to this repo.

**Comment by trxcllnt** [comment]
Filed ops issue here: https://github.com/rapidsai/ops/issues/3069

**Comment by trxcllnt** [comment]
@bdice :tada: looks like both builds are working: [conda](https://github.com/rapidsai/devcontainers/actions/runs/7804789314/job/21287547935?pr=210#step:7:24404), [pip](https://github.com/rapidsai/devcontainers/actions/runs/7804789314/job/21287548140?pr=210#step:7:28979). Now we just gotta make a PR for https://github.com/rapidsai/shared-workflows/tree/fix/devcontainer-json-location.

Ignore the LLVM failure for now, they have an issue with their arm64 dev packages.

File: .github/workflows/test.yml [hunk 12]
```diff
@@ -145,3 +145,29 @@ jobs:
       features: "${{ toJSON(matrix.features) }}"
       container_env: "${{ toJSON(matrix.env) }}"
       repo: "${{ vars.DOCKERHUB_REPOSITORY || github.repository }}"
+
+  build-rapids:
```
Comment (bdice) [inline .github/workflows/test.yml:149]: Do we want to move this to a separate workflow that is triggered nightly before merging?

Comment (trxcllnt) [reply .github/workflows/test.yml:149]: I like having it in test.yaml. We can make it where it's only run if things that affect the RAPIDS projects (like `devcontainer-utils` or `rapids-build-utils` features) are changed.

Comment (bdice) [reply .github/workflows/test.yml:149]: I thought the purpose of this PR would be to have a nightly all-RAPIDS build. Is there a way to add a trigger for that?

Comment (trxcllnt) [reply .github/workflows/test.yml:149]: Sure, we can make it a reusable workflow.

File: .github/workflows/build-all-rapids-repos.yml [hunk 13]
```diff
@@ -29,6 +29,8 @@ jobs:
     secrets: inherit
     uses: rapidsai/shared-workflows/.github/workflows/build-in-devcontainer.yaml@fix/devcontainer-json-location
     with:
+      arch: '["amd64", "arm64"]'
```
[the remark this answers is not in the document]
Comment (trxcllnt) [reply .github/workflows/build-all-rapids-repos.yml:32]: This arch is unrelated to `rapids-dependency-file-generator`.

**Comment by trxcllnt** [comment]
@bdice I'm going to merge this, we can update the shared-workflows devcontainer ref when that PR merges.

**Pull request merged**